packages/astro/src/server/middleware.ts

```diff
@@ -56,107 +56,123 @@ type AstroLocalsWithSentry = Record<string, unknown> & {
   __sentry_wrapped__?: boolean;
 };
 
-export const handleRequest: (options?: MiddlewareOptions) => MiddlewareResponseHandler = (
-  options = { trackClientIp: false, trackHeaders: false },
-) => {
+export const handleRequest: (options?: MiddlewareOptions) => MiddlewareResponseHandler = options => {
+  const handlerOptions = { trackClientIp: false, trackHeaders: false, ...options };
+
   return async (ctx, next) => {
-    // Make sure we don't accidentally double wrap (e.g. user added middleware and integration auto added it)
-    const locals = ctx.locals as AstroLocalsWithSentry;
-    if (locals && locals.__sentry_wrapped__) {
-      return next();
+    // if there is an active span, we know that this handle call is nested and hence
+    // we don't create a new domain for it. If we created one, nested server calls would
+    // create new transactions instead of adding a child span to the currently active span.
+    if (getCurrentHub().getScope().getSpan()) {
+      return instrumentRequest(ctx, next, handlerOptions);
     }
-    addNonEnumerableProperty(locals, '__sentry_wrapped__', true);
+    return runWithAsyncContext(() => {
+      return instrumentRequest(ctx, next, handlerOptions);
+    });
+  };
+};
 
-    const method = ctx.request.method;
-    const headers = ctx.request.headers;
+async function instrumentRequest(
+  ctx: Parameters<MiddlewareResponseHandler>[0],
+  next: Parameters<MiddlewareResponseHandler>[1],
+  options: MiddlewareOptions,
+): Promise<Response> {
+  // Make sure we don't accidentally double wrap (e.g. user added middleware and integration auto added it)
+  const locals = ctx.locals as AstroLocalsWithSentry;
+  if (locals && locals.__sentry_wrapped__) {
+    return next();
+  }
+  addNonEnumerableProperty(locals, '__sentry_wrapped__', true);
 
-    const { dynamicSamplingContext, traceparentData, propagationContext } = tracingContextFromHeaders(
-      headers.get('sentry-trace') || undefined,
-      headers.get('baggage'),
-    );
+  const method = ctx.request.method;
+  const headers = ctx.request.headers;
 
-    const allHeaders: Record<string, string> = {};
-    headers.forEach((value, key) => {
-      allHeaders[key] = value;
-    });
+  const { dynamicSamplingContext, traceparentData, propagationContext } = tracingContextFromHeaders(
```

Comment:
We can now use `continueTrace` for this.

Reply:
good point, will change it!

Reply:
(Great call - avoiding duplicating this logic is worth the refactor)